stdenv/make-derivation: check that all inputs are of an appropriate type

lheckemann opened the pull request:
###### Motivation for this change
Fixes #24462

I'm not sure if this is the best error message we can do; it would be nice to say what the value is, but we don't currently have a function that will briefly and usefully summarise a value, as discussed in the issue. It could be nice to have a mode for `toPretty` that will abbreviate values and could be used for this, but for now I've gone and implemented a quick-win way of getting an already significantly more helpful error message.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

lheckemann committed:
stdenv: check that all inputs are of an appropriate type

Fixes #24462

Ericson2314 commented on pkgs/stdenv/generic/make-derivation.nix:
```diff
@@ -128,6 +128,10 @@ let
+
+  checkDependencyList = name: deps: lib.flip lib.imap1 deps (index: dep:
+    if lib.isDerivation dep || isNull dep || builtins.typeOf dep == "path" then dep
+    else throw "Dependency is not a derivation: element ${toString index} of ${name} for ${attrs.name or attrs.pname}");
```
```suggestion
    else throw "Dependency is not a valid type: element ${toString index} of ${name} for ${attrs.name or attrs.pname}");
```
maybe, since null and paths are in fact supported?

lheckemann replied:
Done :)

Ericson2314 reviewed (approved):
Small nit, but I like the idea.

lheckemann committed:
Merge remote-tracking branch 'upstream/nixos-unstable' into better-error-for-bad-buildinputs

lheckemann commented:
Turns out stdenv allows nested lists for `buildInputs`, and the `nix` package relies on that...

lheckemann committed:
make-derivation: allow nested lists in buildInputs

This isn't really desirable in general, but given that Nix itself
currently relies on this behaviour and that we don't want to break
backwards compatibility we should support it for now, maybe
deprecating it in the future.

lheckemann commented:
Turns out it can be a string as well: https://github.com/NixOS/nixpkgs/blob/15dab8cb52f2d98adc4ad2e4958d3abc1fe0849e/pkgs/development/tools/xcbuild/wrapper.nix#L89

lheckemann committed:
make-derivation: allow strings in build input lists

lheckemann commented:
@dschrempf no need to apologise, I think this is just about the most appropriate place for your comment :) 

Unfortunately there isn't really a way to do that with Nix as it is; it would require having code-as-data functionality in Nix to inspect the expression that produced the wrong type of thing; or probably major changes to Nix itself so that it can produce a better error message by itself.
